### DISCUSSION TITLE
No checksum for conditional dependencies?

### DISCUSSION THREAD

**Discussion by niklaskorz** [yarnpkg/berry#6441] [Q&A] [answered]
I'll first open this as a discussion before opening an issue, so I can figure out of this is intentional behavior or a bug. Conditional dependencies that have more than one condition seem to be lacking a checksum in the lockfile.

This can be observed in yarn's own lockfile, e.g. here with `os=darwin & cpu=arm64`:

https://github.com/yarnpkg/berry/blob/13d5b3041794c33171808fdce635461ff4ab5c4e/yarn.lock#L9719-L9724

Every entry with multiple condition rules lacks a checksum.

Meanwhile, the one conditional dependency here that has only a single condition `os=darwin` has a checksum:

https://github.com/yarnpkg/berry/blob/13d5b3041794c33171808fdce635461ff4ab5c4e/yarn.lock#L10981-L10989

Even if this is intentional behavior, what's the reasoning behind it? This make it harder to map yarn to graph-based isolated build environments like Nix.

**Comment by arcanis** [comment, marked as answer]
> Even if this is intentional behavior, what's the reasoning behind it? This make it harder to map yarn to graph-based isolated build environments like Nix.

Since they are not downloaded, we can't compute their checksum. We can still lock their versions though so they are the same across all environments.

**Reply by niklaskorz** [reply]
Thank your for the explanation! I also now found the origin of this:

https://github.com/yarnpkg/berry/blob/13d5b3041794c33171808fdce635461ff4ab5c4e/packages/yarnpkg-core/sources/Cache.ts#L490-L493

This is somewhat problematic as the hash needs to be known for any network access in Nix, but I think I found a workaround for now. We're using https://github.com/madjam002/yarnpnp2nix, which is a Yarn plugin, so it has direct access to Yarn's store functions, including those for getting checksums of cached files. By setting `supportedArchitectures` to include all configurations that we need in our build environment, we can get the checksums missing in the lockfile for the cached files. This might not be interesting for you personally, but I'll put it here in case anyone else runs into this problem.
